### ISSUE TITLE
Allow for efficient and correct SSA simulations

### ISSUE THREAD

**Issue by sebapersson** [sebapersson/SBMLImporter.jl#18]
Currently when importing by default we set `only_use_rate=true` for a `Catalyst.Reaction`. The drawback with this is that it does not allow for efficient SSA simulations. To allow for efficient SSA simulations for reactions where it makes sense `only_use_rate` should be set to `false`, and for reaction that make sense I mean reaction with:

* Involved species have `only_substance_units=true` (species are given in amount)
* The propensity does not depend on assignment rules, rate-rules etc...

How to do this in practice? If these criteria are fulfilled it should be enough to just divide with the reactants in the propensity (@TorkelE)? To test correctness we can use test-case 1-39 in the stochastic part of the SBML test-suite.

If a user tries to create a `JumpProblem` for a model where `only_substance_units=true` (e.g. that might happen a specie is given in unit of conc.), once the import struct is done #16 , a warning can be thrown in the overload method.

**Comment by TorkelE** [comment]
Catalyst have a function, `oderatelaw(rx::Reaction)`, which computes the propensity of a reaction (rate multiplied by an expression of the substrates, we need to use this one. The problem is that we want to divide by just the substrate expression, which it does not give us, so we have to do some little algebra:
```
p1 = r*s_exp      # The propensity occurring as a rate in the SBML reaction
p2 = r*s_exp^2  # The propensity yielded by applying oderatelaw on an SBML reaction with only_use_rate=false
```
To get the rate we want we need (p1^2)/p2.

I have created a function, `updated_rate_reaction`, which takes a reaction with propensity of the SBML sort (where `only_use_rate=false` was used), and returns a reaction with the rate we want:
```julia
using Catalyst
import Setfield: @set

function updated_rate_reaction(rx)
    @set rx.rate = (rx.rate^2) / Catalyst.oderatelaw(rx)
end
```
We can apply these to a variety of reactions (on the form SBML would generate, the comment implies the real rate, which we want to achieve).
```julia
rx1 = @reaction k, 0 --> X              # k
rx2 = @reaction k*X, X --> 0            # k
rx3 = @reaction k*X*Y, X + Y --> 2X     # k
rx4 = @reaction k*(X^3)/6, 3X --> X     # k
rx5 = @reaction hill(X,v,K,n), 0 --> X  # hill(X,v,K,n)
rx6 = @reaction mm(X,v,K)*X, X --> Y    # mm(X,v,K)
rx7 = @reaction k*(X^2)*Y*Z/2, 2X + Y --> X # k*Z     

updated_rate_reaction(rx1)
updated_rate_reaction(rx2)
updated_rate_reaction(rx3)
updated_rate_reaction(rx4)
updated_rate_reaction(rx5)
updated_rate_reaction(rx6)
updated_rate_reaction(rx7)
```
This depends on Symbolics automatically simplifying the expressions. If that is not the case, we might have a problem (And maybe model with really convoluted rate which is hard to simplify for symbolic engines is part of the SBML suite?). We should still get something which is correct though (but with an unnecessarily complicated rate to evaluate).

Basically, in SBML importer, wherever you have `Reaction(...)`, this can just be replaced with `updated_rate_reaction(Reaction(...))`

### One potential problem (that can be dealt with)
In Catalyst, when non-unitary stoichiometric occur, we scale these combinatorial (which should be done, but many people don't do). E.g. for
```julia
@reaction k, 2X + 3Y --> 0
```
the propensity is not `k*(X^2)*(Y^3)`, but `k*(X^2)/2!*(Y^3)/3!` = `k*(X^2)*(Y^3)/12`. There is an option in `oderatelaw` to disable this, which fixes things. So depending on what SBML do here, we might to use this one. The annoying case is if this is inconsistent, and varied between different models.

**Comment by sebapersson** [comment]
This looks good, I will test the `updated_rate_reaction` function. Regarding the comment on whether or not Symbolics will be able to simplify the expression correct, for those ODE models were stochastic SSA simulations are reasonable (e.g. units not in concentration, few if no rate-laws), I do not see that the expressions should be to complex. 

One the potential problem, non-unitary stoichiometric is when the stoichiometry is not equal to 1? Currently for a reaction like 

```julia
@reaction k, 2X + 3Y --> 0
```

SBMLImporter produces the expression `k*(X^2)*(Y^3)`

**Comment by TorkelE** [comment]
Ok, in that case using this version should be enough:
```julia
function updated_rate_reaction(rx)
    @set rx.rate = (rx.rate^2) / Catalyst.oderatelaw(rx; combinatoric_ratelaw=false)
end
```

Then, make `combinatoric_ratelaw=false` a kwarg to your SBML import function, so in case some have properly accounted for this, it can be set to `true` and the model is loaded properly.

**Comment by sebapersson** [comment]
With #19 the code now passes all except 2 stochastic tests (assignment-rule and `ContiniousCallback`). 

The test-cases are quite small (but have analytical solution), so for completeness in testing if you @TorkelE happen to have a bigger SBML model (also implemented in Catalyst) would be convenient.

**Comment by sebapersson** [comment]
With #26  and #19 I consider this closed.
